packages/ckeditor5-clipboard/src/clipboardmarkersutils.ts

```diff
@@ -106,25 +105,127 @@ export default class ClipboardMarkersUtils extends Plugin {
 	 * @internal
 	 */
 	public _copySelectedFragmentWithMarkers(
-		action: ClipboardMarkerAction,
+		action: ClipboardMarkerRestrictedAction,
 		selection: Selection | DocumentSelection,
-		getCopiedFragment: ( writer: Writer ) => DocumentFragment = writer => writer.model.getSelectedContent( selection )
+		getCopiedFragment: ( writer: Writer ) => DocumentFragment = writer =>
+			writer.model.getSelectedContent( writer.model.document.selection )
 	): DocumentFragment {
 		return this.editor.model.change( writer => {
-			const sourceSelectionInsertedMarkers = this._insertFakeMarkersToSelection( writer, selection, action );
+			const oldSelection = writer.model.document.selection;
+
+			// In some scenarios, such like in drag & drop, passed `selection` parameter is not actually
+			// the same `selection` as the `writer.model.document.selection` which means that `_insertFakeMarkersToSelection`
+			// is not affecting passed `selection` `start` and `end` positions but rather modifies `writer.model.document.selection`.
+			//
+			// It is critical due to fact that when we have selection that starts [ 0, 0 ] and ends at [ 1, 0 ]
+			// and after inserting fake marker it will point to such marker instead of new widget position at start: [ 1, 0 ] end: [2, 0 ].
+			// `writer.insert` modifies only original `writer.model.document.selection`.
```

Comment:
In general, this gets me worried. I don't get what happens in drag and drop and why we need to fix it, but if we try hard to operate on document selection all the time, and then there is a mechanism that operates on arbitrary selection, and we are doing some fixing and still try to operate on document selection, I am afraid that at some point this will get mismatched.

Maybe after refactor we will not have this problem anymore.